mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,279 @@
+                RetTypes.push_back(v.getType());
+            }
+        }
+
+        // We process the tapes in reverse order
+        // Hence from the viewpoint of the earlier tapes,
+        // a processed later tape (which has become a call op) is also their "post processing"
+        // The later tape's arguments need to be returned by the earlier tapes
+        // Hence when processing a tape, it needs to ask for earlier tapes to return
+        // values it needs, just like how post processing asks for all tapes to return
+        // values it needs.
+        for (Operation *op : *TapeOps) {
+            for (auto operand : op->getOperands()) {
+                Operation *OperandSource = operand.getDefiningOp();
+                if (std::find(TapeOps->begin(), TapeOps->end(), OperandSource) == TapeOps->end()) {
+                    // A tape operand not produced in this tape itself,
+                    // must be from the tapes/preprocessing!
+                    // Need to be replaced by the previous tape functions' return values
+                    NecessaryValuesFromEarlierTapes.push_back(operand);
+                }
+            }
+        }
+
+        // 2. Create a scf.execute_region and wrap it around the tape ops
+        // The scf.execute_region needs to yield (aka return) the values identified
+        builder.setInsertionPoint(TapeOps->front());
+        scf::ExecuteRegionOp executeRegionOp = builder.create<scf::ExecuteRegionOp>(
+            OriginalMultitapeFunc->getLoc(), ArrayRef(RetTypes));
+
+        builder.setInsertionPointToStart(&executeRegionOp.getRegion().emplaceBlock());
+        mlir::Block::iterator it = executeRegionOp.getRegion().front().end();
+        for (auto op : *TapeOps) {
+            op->moveBefore(&executeRegionOp.getRegion().front(), it);
+        }
+
+        builder.setInsertionPointAfter(&executeRegionOp.getRegion().front().back());
+        Operation *y =
+            builder.create<scf::YieldOp>(OriginalMultitapeFunc->getLoc(), ArrayRef(RetValues));
+
+        // 3. The scf outliner automatically captures values not defined in the
+        // to-be-outlined function body and transforms them to the outlined function's
+        // arguments, but it does not automatically propagate the returned value
+        // downstream to replace the previously used values. This we need to do manually.
+
+        // At this stage, the scf.yield values ARE the ones to be replaced by
+        // the scf region's result values, since we never replaced anything yet
+        // Of course, don't replace the use in the region itself!
+
+        SmallPtrSet<Operation *, 16> exceptions;
+        exceptions.insert(y);
+        for (auto op : *TapeOps) {
+            exceptions.insert(op);
+        }
+
+        for (size_t i = 0; i < RetValues.size(); i++) {
+            RetValues[i].replaceAllUsesExcept(executeRegionOp->getResults()[i], exceptions);
+        }
+
+        // 4. Outline the region
```

Comment:
Can you outline this into its own function?

Reply:
I think this last step makes more sense to not be outlined, since it's just one line that calls something else?